packages/components/src/form-token-field/index.tsx

```diff
@@ -0,0 +1,696 @@
+		placeholder,
+		label = __( 'Add item' ),
+		className,
+		suggestions = [],
+		maxSuggestions = 100,
+		value = [],
+		displayTransform = identity,
+		saveTransform = ( token ) => token.trim(),
+		onChange = () => {},
+		onInputChange = () => {},
+		onFocus = undefined,
+		isBorderless = false,
+		disabled = false,
+		tokenizeOnSpace = false,
+		messages = {
+			added: __( 'Item added.' ),
+			removed: __( 'Item removed.' ),
+			remove: __( 'Remove item' ),
+			__experimentalInvalid: __( 'Invalid item' ),
+		},
+		__experimentalExpandOnFocus = false,
+		__experimentalValidateInput = () => true,
+		__experimentalShowHowTo = true,
+	} = props;
+
+	const instanceId = useInstanceId( FormTokenField );
+
+	const [ incompleteTokenValue, setIncompleteTokenValue ] = useState( '' );
+	const [ inputOffsetFromEnd, setInputOffsetFromEnd ] = useState( 0 );
+	const [ isActive, setIsActive ] = useState( false );
+	const [ isExpanded, setIsExpanded ] = useState( false );
+	const [ selectedSuggestionIndex, setSelectedSuggestionIndex ] = useState(
+		-1
+	);
+	const [ selectedSuggestionScroll, setSelectedSuggestionScroll ] = useState(
+		false
+	);
+
+	const debouncedSpeak = useDebounce( speak, 500 );
+	const prevSuggestions = usePrevious< string[] >( suggestions );
+	const input = useRef< HTMLInputElement >( null );
+	const tokensAndInput = useRef< HTMLInputElement >( null );
+
+	useEffect( () => {
+		// Make sure to focus the input when the isActive state is true.
+		if ( isActive && ! hasFocus() ) {
+			focus();
+		}
+	}, [ isActive ] );
+
+	useEffect( () => {
+		const suggestionsDidUpdate = ! isShallowEqual(
+			suggestions,
+			prevSuggestions || []
+		);
+
+		updateSuggestions( suggestionsDidUpdate );
+	}, [ suggestions ] );
+
+	useEffect( () => {
+		updateSuggestions();
+	}, [ incompleteTokenValue ] );
+
+	if ( disabled && isActive ) {
+		setIsActive( false );
+		setIncompleteTokenValue( '' );
+	}
+
+	function focus() {
+		input.current?.focus();
+	}
+
+	function hasFocus() {
+		return input.current === input.current?.ownerDocument.activeElement;
+	}
+
+	function onFocusEventHandler( event: FocusEvent ) {
+		// If focus is on the input or on the container, set the isActive state to true.
+		if ( hasFocus() || event.target === tokensAndInput.current ) {
+			setIsActive( true );
+			setIsExpanded( __experimentalExpandOnFocus || isExpanded );
+		} else {
+			/*
+			 * Otherwise, focus is on one of the token "remove" buttons and we
+			 * set the isActive state to false to prevent the input to be
+			 * re-focused, see componentDidUpdate().
+			 */
+			setIsActive( false );
+		}
+
+		if ( 'function' === typeof onFocus ) {
+			onFocus( event );
+		}
+	}
+
+	function onBlur() {
+		if ( inputHasValidValue() ) {
+			setIsActive( false );
+		} else {
+			setIncompleteTokenValue( '' );
+			setInputOffsetFromEnd( 0 );
+			setIsActive( false );
+			setIsExpanded( false );
+			setSelectedSuggestionIndex( -1 );
+			setSelectedSuggestionScroll( false );
+		}
```

Comment:
Should we perhaps add a comment here?

```suggestion
		} else {
			// Reset to initial state
			setIncompleteTokenValue( '' );
			setInputOffsetFromEnd( 0 );
			setIsActive( false );
			setIsExpanded( false );
			setSelectedSuggestionIndex( -1 );
			setSelectedSuggestionScroll( false );
		}
```